### PR TITLE
feat: implement REST service for extension registry

### PR DESCRIPTION
**Pull request by stefannica**
* add operational attributes (created/updated) to extension,
service and credentials elements
* add GetAllExtensions method to extension registry
* fix the interface signature for the extension registry and store
* add goa design for extension registry
* implement REST service for extension registry

Implements: https://github.com/fuseml/fuseml/issues/74

### CODE REVIEW

File: pkg/core/extension_store.go [hunk 1]
```diff
@@ -493,6 +516,8 @@ func (store *ExtensionStore) UpdateExtension(ctx context.Context, extension *dom
 	if err != nil {
 		return err
 	}
+	extension.Registered = extRecord.Registered
```
Comment (jsuchome) [inline pkg/core/extension_store.go:519]: Why `Registered`?

File: pkg/core/extension_store.go [hunk 2]
```diff
@@ -503,6 +528,8 @@ func (store *ExtensionStore) UpdateService(ctx context.Context, service *domain.
 	if err != nil {
 		return err
 	}
+	service.Registered = svcRecord.Registered
```
Comment (jsuchome) [inline pkg/core/extension_store.go:531]: Why?

File: pkg/core/extension_store.go [hunk 3]
```diff
@@ -523,6 +550,8 @@ func (store *ExtensionStore) UpdateCredentials(ctx context.Context, credentials
 	if err != nil {
 		return err
 	}
+	credentials.Created = credRecord.Created
```
Comment (jsuchome) [inline pkg/core/extension_store.go:553]: Again, why even touch Created on Update operation?

File: design/extension.go [hunk 4]
```diff
@@ -0,0 +1,963 @@
...
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("NotFound", CodeNotFound)
+			Response("BadRequest", CodeInvalidArgument)
+			Response("Conflict", CodeAlreadyExists)
+		})
+	})
+
+	Method("getService", func() {
```
Comment (jsuchome) [inline design/extension.go:180]: Shouldn't `getService` be just a special case of `listServices`, with `id` argument being optional? I think we have it similar for other resources

File: design/extension.go [hunk 5]
```diff
@@ -0,0 +1,963 @@
...
+			Response("NotFound", CodeNotFound)
+			Response("BadRequest", CodeInvalidArgument)
+		})
+	})
+
+})
+
+// Extension descriptor
+var Extension = Type("Extension", func() {
+	tag := 1
```
Comment (jsuchome) [inline design/extension.go:661]: eh, what's the tag for? I see it for the first time..

File: design/extension.go [hunk 6]
```diff
@@ -0,0 +1,963 @@
...
+	})
+	tag++
+	Field(tag, "service_id", String, "Reference to the service this endpoint belongs to", func() {
+		Pattern(identifierPattern)
+		Default("")
+		MaxLength(100)
+		Example("s3")
+	})
+	tag++
+	Field(tag, "endpoint_type", String,
```
Comment (jsuchome) [inline design/extension.go:851]: This should be just `"type"`, right?

File: pkg/core/extension_store.go [hunk 7]
```diff
@@ -292,7 +292,7 @@ func (store *ExtensionStore) getExtensionRecord(ctx context.Context, extensionID
 
 // GetAllExtensionss - retrieve all registered extensions, with all participating services, endpoints and credentials
 func (store *ExtensionStore) GetAllExtensions(ctx context.Context) (result []*domain.ExtensionRecord, err error) {
-	result = make([]*domain.ExtensionRecord, len(store.items))
+	result = make([]*domain.ExtensionRecord, 0)
```
Comment (jsuchome) [inline pkg/core/extension_store.go:295]: Why was this change needed?

File: design/extension.go [hunk 8]
```diff
@@ -0,0 +1,963 @@
...
+
+		GRPC(func() {
+			Response(CodeOK)
+		})
+	})
+
+	Method("updateExtension", func() {
+		Description("Update an extension registered in FuseML")
+
+		Payload(Extension, "Extension update request")
```
Comment (jsuchome) [inline design/extension.go:88]: I understand that you get the extension ID from the extension struct passed as parameter, but I am still not sure about this.

Isn't it common to pass the ID explicitly during the update operations, so that the object to update is clearly identified and only the _other_ parameter holds the data which should be used for the actual update?

What if, for example, you want to change the ID itself? Your methods will not find the original objects...

File: pkg/core/extension_store.go (hunk 1 above)
Comment (stefannica) [reply pkg/core/extension_store.go:519]: This is to avoid overwriting the registered timestamp when the extension is updated

File: pkg/core/extension_store.go (hunk 2 above)
Comment (stefannica) [reply pkg/core/extension_store.go:531]: Same as before: the registered timestamp is immutable for the lifetime of an extension/service/set of credentials.

File: pkg/core/extension_store.go (hunk 3 above)
Comment (stefannica) [reply pkg/core/extension_store.go:553]: The next operation overwrites everything in the in-memory database with the content supplied as input.

File: pkg/core/extension_store.go (hunk 7 above)
Comment (stefannica) [reply pkg/core/extension_store.go:295]: There was previously no method to allow listing all extensions in the registry/store. This is needed to implement GET /extensions

File: design/extension.go (hunk 4 above)
Comment (stefannica) [reply design/extension.go:180]: No, they are two different things. getService targets a specific service and returns an error if the service isn't found. listServices would only return an empty list if the criteria isn't matching any of the services.

These REST operations also target different REST resources (look at the URL route), which is one more reason why they should be different commands.

File: design/extension.go (hunk 5 above)
Comment (stefannica) [reply design/extension.go:661]: gRPC fields require a unique tag for each attribute

File: pkg/core/extension_store.go (hunk 7 above)
Comment (jsuchome) [reply pkg/core/extension_store.go:295]: Ah, this comment might be confusing. I was referring to the change between first and second commit:
```
-	result = make([]*domain.ExtensionRecord, len(store.items))
+	result = make([]*domain.ExtensionRecord, 0)
```

File: design/extension.go (hunk 6 above)
Comment (stefannica) [reply design/extension.go:851]: `type` is a reserved word in golang. I would rather avoid triggering it and just use this name instead. But yes, should be just `type`.

Comment (jsuchome) [reply design/extension.go:851]: Well, you are using `type` in the yaml example for credentials...

File: design/extension.go (hunk 4 above)
Comment (jsuchome) [reply design/extension.go:180]: Ok

File: pkg/core/extension_store.go (hunk 7 above)
Comment (stefannica) [reply pkg/core/extension_store.go:295]: Oh, I see. Well, you either initialize the slice to the intended size and use indexing (e.g. `result[i] = value`) or you initialize it to zero size and use `append`. In this case, I fixed a bug that gave me a bunch of `nil` values as the first entries.

File: design/extension.go (hunk 8 above)
Comment (stefannica) [reply design/extension.go:88]: I've been battling with this same concept for a couple of days now. Yes, you're right, they should be different things: the ID should be a URL parameter or part of the URL path, the rest of the attributes should be in the request body.

Given that the extension registry implementation doesn't allow changing the ID attribute during an update (and I don't intend to add support for that because it complicates things unnecessarily), I thought reusing the body attributes would provide for a cleaner implementation, which it does.

I can try switching to use parameters _in addition_ to the ID attributes in the body. The one thing I can't do is to remove the ID attributes from the body, because that would mean having to create another set of objects like Extension and ExtensionService that don't include those IDs, which would mean having to duplicate all the boilerplate code that is required to convert them to/from domain.

File: design/extension.go (hunk 6 above)
Comment (stefannica) [reply design/extension.go]: Ok, done

File: pkg/core/extension_store.go (hunk 1 above)
Comment (jsuchome) [reply pkg/core/extension_store.go:519]: I see. The `Registered` field is generated and we can't expect user to provide it on update.

But this brings another question - what if user wants to update just one entry of the object (say, Description), and provides only this in the POST method. With current implementation, all other fields will be deleted, right? Should this be the intended behavior of the  `Update` methods?

File: design/extension.go (hunk 8 above)
Comment (stefannica) [reply design/extension.go:88]: Ok, it looks like I need the IDs to be in the payload too, even if they're modeled as parameters. Besides, I don't need to model them as parameters because they are part of the URL path anyway.

This only leaves me with the idea of having to separate the IDs from the regular payload, which means the request body will have to look like this:

```
# curl -X PUT --data-binary @update-credentials001.yaml -H "Content-Type: application/x-yaml" http://localhost:8000/extensions/mlflow-0001/services/mlflow-store/credentials/default-s3-account
credentials:
  scope: project
  projects:
    - project-001
  configuration:
    AWS_ACCESS_KEY_ID: oT0SfbJPEu6kUbUKsH24
    AWS_SECRET_ACCESS_KEY: GiZff8KqS5xWQ4eagRujh1tDcbQyRP0bEJSBOfs5
```

I really don't see that this will make a difference. The user could still include the IDs in the update request, e.g.:

```
id: default-s3-account
extension_id: mlflow-0001
service_id: mlflow-store
credentials:
  scope: project
  projects:
    - project-001
  configuration:
    AWS_ACCESS_KEY_ID: oT0SfbJPEu6kUbUKsH24
    AWS_SECRET_ACCESS_KEY: GiZff8KqS5xWQ4eagRujh1tDcbQyRP0bEJSBOfs5
```

At least with the current approach, the IDs that are included in the request are automatically overwritten by the REST service, so I don't need to worry about complications.

I would like to stick with the current implementation, if that's the same to you.

File: pkg/core/extension_store.go (hunk 7 above)
Comment (jsuchome) [reply pkg/core/extension_store.go:295]: What about initializing it with 0 size but `len(store.items)` capacity? You could still use `append` there.

File: design/extension.go (hunk 8 above)
Comment (jsuchome) [reply design/extension.go:88]: Hm, I think the approach with identifiers being separate (not necessary extracted from the payload) is nicer, but I guess I do not really mind.

Anyway, please take a look at the other Update related issue which I mention here https://github.com/fuseml/fuseml-core/pull/93#discussion_r691183901

File: pkg/core/extension_store.go (hunk 1 above)
Comment (stefannica) [reply pkg/core/extension_store.go:519]: Good question.

To implement this type of surgical update, we need to make all the fields in the design objects optional (already are) and also to not give them a default value. This will force goa to generate pointers as attributes, instead of actual values as they are now.

The other solution is to push this mechanism further out, into the CLI and UI (when available).

Changing everything to pointers now is going to be a PITA to refactor, but I'll give it a shot.

File: pkg/core/extension_store.go (hunk 7 above)
Comment (jsuchome) [reply pkg/core/extension_store.go:295]: (but that's not necessary to change)

Comment (stefannica) [reply pkg/core/extension_store.go]: done

File: pkg/core/extension_store.go (hunk 1 above)
Comment (stefannica) [reply pkg/core/extension_store.go:519]: Done. Was easier than I thought.

**Review by jsuchome — APPROVED** [review]
Good to go
